blocks/editable/index.js

```diff
@@ -68,18 +68,26 @@ export default class Editable extends Component {
 		this.onSelectionChange = this.onSelectionChange.bind( this );
 		this.maybePropagateUndo = this.maybePropagateUndo.bind( this );
 		this.onPastePostProcess = this.onPastePostProcess.bind( this );
+		this.onToolbarAddLink = this.onToolbarAddLink.bind( this );
+		this.onToolbarEditLink = this.onToolbarEditLink.bind( this );
+		this.onToolbarChangeLinkValue = this.onToolbarChangeLinkValue.bind( this );
 
 		this.state = {
 			formats: {},
-			bookmark: null,
 			empty: ! props.value || ! props.value.length,
+			toolbar: {
+				isAddingLink: false,
+				isEditingLink: false,
+				newLinkValue: '',
+			},
```

Comment:
These all need to be reset in `onNodeChange` which `FormatTooblar` has no knowledge of.  This also removes the need for trying to reconcile state in `componentWillReceiveProps` and the `setTimeout` hack that was in `addLink`.  I can't think of a cleaner way to do this but happy to take suggestions!